test_runner/regress/test_compatibility.py

```diff
@@ -552,3 +584,84 @@ def test_historic_storage_formats(
     env.pageserver.http_client().timeline_compact(
         dataset.tenant_id, existing_timeline_id, force_image_layer_creation=True
     )
+
+
+@check_ondisk_data_compatibility_if_enabled
+@pytest.mark.xdist_group("compatibility")
+@pytest.mark.parametrize("combination", [0x15, 0x02, 0x0C, 0x09, 0x1A])
+def test_versions_mismatch(
+    neon_env_builder: NeonEnvBuilder, test_output_dir: Path, pg_version: PgVersion, combination
+):
+    # Combination: 0: old, 1: new
+    bitmap = {
+        "storage_controller": 0x01,
+        "storage_broker": 0x02,
+        "compute": 0x04,
+        "safekeeper": 0x08,
+        "pageserver": 0x10,
+    }
+
+    binaries = {
+        "storage_controller": ["storage_controller"],
+        "storage_broker": ["storage_broker"],
+        "compute": ["compute_ctl"],
+        "safekeeper": ["safekeeper"],
+        "pageserver": ["pageserver", "pagectl"],
+    }
+
+    compatibility_snapshot_dir_env = os.environ.get("COMPATIBILITY_SNAPSHOT_DIR")
+    assert compatibility_snapshot_dir_env is not None, (
+        f"COMPATIBILITY_SNAPSHOT_DIR is not set. It should be set to `compatibility_snapshot_pg{pg_version.v_prefixed}`"
+        "path generateted by test_create_snapshot (ideally generated by the previous version of Neon)"
+    )
+    compatibility_snapshot_dir = Path(compatibility_snapshot_dir_env).resolve()
+
+    postgres_distrib_dir_env = os.environ.get("POSTGRES_DISTRIB_DIR")
+    assert (
+        postgres_distrib_dir_env is not None
+    ), "POSTGRES_DISTRIB_DIR is not set. It should be set to a path with postgres binaries"
+    postgres_distrib_dir = Path(postgres_distrib_dir_env).resolve()
+
+    compatibility_postgres_distrib_dir_env = os.environ.get("COMPATIBILITY_POSTGRES_DISTRIB_DIR")
+    assert compatibility_postgres_distrib_dir_env is not None, (
+        "COMPATIBILITY_POSTGRES_DISTRIB_DIR is not set. It should be set to a path with postgres binaries"
+        "(ideally generated by the previous version of Neon)"
+    )
+    compatibility_postgres_distrib_dir = Path(compatibility_postgres_distrib_dir_env).resolve()
+
+    neon_bin_env = os.environ.get("NEON_BIN")
+    assert (
+        neon_bin_env is not None
+    ), "NEON_BIN is not set. It should be set to a path with Neon binaries "
+    neon_bin = Path(neon_bin_env).resolve()
+
+    compatibility_neon_bin_env = os.environ.get("COMPATIBILITY_NEON_BIN")
+    assert compatibility_neon_bin_env is not None, (
+        "COMPATIBILITY_NEON_BIN is not set. It should be set to a path with Neon binaries "
+        "(ideally generated by the previous version of Neon)"
+    )
+    compatibility_neon_bin = Path(compatibility_neon_bin_env).resolve()
+
+    target_dir = test_output_dir / "neon" / "bin"
+    os.makedirs(target_dir)
```

Comment:
I'd try not to mix `pathlib`'s API with `os`' counterparts, let's use [`pathlib.Path.mkdir`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir) instead, it should be something like:
```python
target_dir.mkdir(parents=True)
```